common/src/main/java/com/wynntils/models/lootrun/LootrunModel.java

```diff
@@ -222,26 +308,155 @@ private void handleStateChange(LootrunningState oldState, LootrunningState newSt
-            WynntilsMod.info("Selected a " + closestBeacon.color() + " beacon at " + closestBeacon.location());
+            WynntilsMod.info("Selected a " + closestBeacon.color() + " beacon at " + closestBeacon.position());
             selectedBeacons.put(closestBeacon.color(), selectedBeacons.getOrDefault(closestBeacon.color(), 0) + 1);
             selectedBeaconsStorage.touched();
             WynntilsMod.postEvent(new LootrunBeaconSelectedEvent(closestBeacon));
+
+            // We selected a beacon, so other beacons are no longer relevant.
+            beacons.clear();
             setClosestBeacon(null);
+            LOOTRUN_BEACON_COMPASS_PROVIDER.reloadTaskMarkers();
+
+            return;
+        }
+    }
+
+    private void updateTaskLocationPrediction(Beacon beacon) {
+        Optional<LootrunLocation> location = getLocation();
+        if (location.isEmpty()) {
+            WynntilsMod.warn("Location was when trying to predict for: " + beacon);
+            return;
+        }
+
+        Set<TaskLocation> currentTaskLocations = taskLocations.get(location.get());
+        if (currentTaskLocations == null || currentTaskLocations.isEmpty()) {
+            WynntilsMod.warn("No task locations found for " + location.get());
             return;
         }
+
+        List<TaskPrediction> usedTaskLocations = beacons.entrySet().stream()
+                .filter(entry -> entry.getKey() != beacon.color())
+                .map(Map.Entry::getValue)
+                .toList();
+
+        Map<Double, TaskLocation> predictionScores = new TreeMap<>();
+        for (TaskLocation currentTaskLocation : currentTaskLocations) {
+            Pair<Double, TaskLocation> prediction = calculatePredictionScore(beacon, currentTaskLocation);
+            if (prediction == null) continue;
+
+            predictionScores.put(prediction.a(), prediction.b());
+        }
+
+        // According to TreeMap's sort order, the first entry is the lowest prediction score.
+        for (Map.Entry<Double, TaskLocation> entry : predictionScores.entrySet()) {
+            TaskLocation closestTaskLocation = entry.getValue();
+            Double predictionValue = entry.getKey();
+
+            TaskPrediction oldPrediction = beacons.get(beacon.color());
+            TaskPrediction newTaskPrediction = new TaskPrediction(beacon, closestTaskLocation, predictionValue);
+
+            // If the prediction is the same, don't update.
+            if (oldPrediction != null
+                    && Objects.equals(oldPrediction.taskLocation(), newTaskPrediction.taskLocation())) {
+                if (newTaskPrediction.predictionScore() < oldPrediction.predictionScore()) {
+                    // The prediction is the same, but the score is better, so update.
+                    beacons.put(beacon.color(), newTaskPrediction);
+                }
+                return;
+            }
+
+            // The prediction is a location that an other color beacon is already at
```

Comment:
Grammar is a bit odd here, I don't fully understand this comment. Can you try to rephrase it?